moda-contracts/src/Marketplace.sol

```diff
@@ -0,0 +1,236 @@
+        if (treasuryFee_ == 0) revert TreasuryFeeCannotBeZero();
+        if (treasury_ == address(0)) revert CannotBeZeroAddress();
+        treasury = treasury_;
+        treasuryFee = treasuryFee_;
+        _usdc = usdc;
+        _catalog = catalog;
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    // External Functions
+
+    /**
+     * @inheritdoc IMarketplace
+     */
+    function createSale(
+        address payable beneficiary,
+        address releases,
+        uint256 tokenId,
+        uint256 amountTotal,
+        uint256 pricePerToken,
+        uint256 startAt,
+        uint256 endAt,
+        uint256 maxCountPerWallet
+    ) external {
+        if (IReleasesRegistration(_catalog).getReleasesOwner(releases) == address(0)) {
+            revert ReleasesIsNotRegistered();
+        }
+        if (beneficiary == address(0)) revert CannotBeZeroAddress();
+        if (amountTotal == 0) revert TokenAmountCannotBeZero();
+        if (endAt != 0 && startAt > endAt) {
+            revert StartCannotBeAfterEnd(startAt, endAt);
+        }
+        if (maxCountPerWallet == 0) revert MaxCountCannotBeZero();
+        IERC1155(releases).safeTransferFrom(_msgSender(), address(this), tokenId, amountTotal, "");
+
+        sales[_msgSender()].push(
+            Sale({
+                releaseOwner: _msgSender(),
+                beneficiary: beneficiary,
+                releases: releases,
+                tokenId: tokenId,
+                amountRemaining: amountTotal,
+                amountTotal: amountTotal,
+                pricePerToken: pricePerToken,
+                startAt: startAt,
+                endAt: endAt,
+                maxCountPerWallet: maxCountPerWallet
+            })
+        );
+
+        emit SaleCreated(_msgSender(), sales[_msgSender()].length - 1);
+    }
+
+    /**
+     * @inheritdoc IMarketplace
+     */
+    function purchase(
+        address seller,
+        uint256 saleId,
+        uint256 tokenAmount,
+        address recipient
+    ) external nonReentrant {
+        Sale storage sale = _getSaleForPurchase(seller, saleId, tokenAmount);
+
+        uint256 totalPrice = sale.pricePerToken * tokenAmount;
+        uint256 fee = (treasuryFee * totalPrice) / 10_000;
+        IERC20(_usdc).safeTransferFrom(_msgSender(), address(this), totalPrice);
+        IERC20(_usdc).transfer(treasury, fee);
+        IERC20(_usdc).transfer(sale.beneficiary, totalPrice - fee);
+
+        _transferTokens(sale.releases, sale.tokenId, tokenAmount, recipient);
+
+        sale.amountRemaining -= tokenAmount;
+
+        emit Purchase(
+            sale.releases, sale.tokenId, recipient, seller, saleId, tokenAmount, block.timestamp
+        );
+    }
+
+    /**
+     * @inheritdoc IMarketplace
+     */
+    function withdraw(uint256 saleId, uint256 tokenAmount) external nonReentrant {
+        if (tokenAmount == 0) revert TokenAmountCannotBeZero();
+        Sale storage sale = sales[_msgSender()][saleId];
```

Comment:
We have to think about how a manager can withdraw for an artist. We want to withdraw for a releaseOwner, not the seller. 

When we transfer we want to transfer back to the Releases contract or the owner, not the manager who may be the one calling this method.